src/index/addrindex.cpp

```diff
@@ -0,0 +1,263 @@
+
+    std::unique_ptr<CDBIterator> iter(NewIterator());
+    iter->Seek(search_key);
+    while (iter->Valid()) {
+        DBKey key;
+        DBValue value;
+        if (!iter->GetKey(key) || key.m_addr_id != addr_id || !iter->GetValue(value) ) break;
+
+        // Check that the stored script matches the one we're searching for, in case of hash collisions.
+        if (value.second != script) continue;
+
+        result.emplace_back(std::make_pair(key, value));
+        iter->Next();
+    }
+
+    return result;
+}
+
+bool AddrIndex::Init() {
+        m_hash_seed = m_db->SetupHashSeed();
+        return BaseIndex::Init();
+}
+
+AddrIndex::AddrIndex(size_t n_cache_size, bool f_memory, bool f_wipe)
+    : m_db(MakeUnique<AddrIndex::DB>(n_cache_size, f_memory, f_wipe)) {}
+
+unsigned int AddrIndex::DB::SetupHashSeed() {
+    static const auto seed_key = std::make_pair(DB_ADDR_INDEX, static_cast<uint8_t>(DBKeyType::SEED));
+    unsigned int seed;
+
+    std::unique_ptr<CDBIterator> iter(NewIterator());
+    std::pair<char, uint8_t> key;
+
+    // If key is in the index already, read it and return.
+    iter->Seek(seed_key);
+    if (iter->Valid() && iter->GetKey(key) && key == seed_key && iter->GetValue(seed)) {
+        return seed;
+    }
+
+    // Generate a random key and write it to the index.
+    seed = GetRandInt(std::numeric_limits<int>::max());
+    Write(seed_key, seed);
+    return seed;
+}
+
+AddrIndex::~AddrIndex() {}
+
+unsigned int AddrIndex::GetAddrId(const CScript& script) {
+    std::vector<unsigned char> script_data;
+    for (auto it = script.begin(); it != script.end(); ++it) {
+        script_data.push_back(*it);
+    }
+    return MurmurHash3(m_hash_seed, script_data);
+}
+
+bool AddrIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
+{
+    CBlockUndo block_undo;
+    CDiskTxPos pos(pindex->GetBlockPos(), GetSizeOfCompactSize(block.vtx.size()));
+    std::vector<std::pair<DBKey, DBValue>> entries;
+
+    const bool not_genesis_block = (pindex->nHeight > 0);
+    if (not_genesis_block && !UndoReadFromDisk(block_undo, pindex)) {
+      return false;
+    }
+
+    for (size_t i = 0; i < block.vtx.size(); ++i) {
+        const CTransaction& tx = *(block.vtx[i]);
+        const uint256 tx_hash = tx.GetHash();
+        for (size_t j = 0; j < tx.vout.size(); ++j) {
+            CScript script_pub_key = tx.vout[j].scriptPubKey;
+            DBKey key(DBKeyType::CREATED, GetAddrId(script_pub_key), COutPoint(tx_hash, j));
+            entries.emplace_back(key, std::make_pair(pos, script_pub_key));
+        }
+
+        // Skip coinbase inputs.
+        if (not_genesis_block && i > 0) {
+            const CTxUndo& tx_undo = block_undo.vtxundo[i-1];
+            for (size_t k = 0; k < tx.vin.size(); ++k) {
+                CScript spent_outputs_scriptpubkey = tx_undo.vprevout[k].out.scriptPubKey;
+                DBKey key(DBKeyType::SPENT, GetAddrId(spent_outputs_scriptpubkey), tx.vin[k].prevout);
```

Comment:
I *think* this is correct because he is using the prevout in the *key*. When looking up by this script with this outpoint, one would want to find this transaction because it spends this outpoint.

AFAICT the value just contains the position of the relevant transaction on disk and the scriptPubKey (to detect collisions).  The value does not contain any indexes into inputs or outputs.

Reply:
Can you add a comment with your key/value format somewhere and the justification?